File: scala/src/main/scala/tddmicroexercises/leaderboard/Race.scala
package tddmicroexercises.leaderboard

class Race(val name: String, val results: List[Driver]) {
  
  val driverNames: List[String] = results.map( (driver: Driver ) => {
    if (driver.isInstanceOf[SelfDrivingCar]) {
      s"Self Driving Car - ${driver.country} (${driver.asInstanceOf[SelfDrivingCar].algorithmVersion})"
    } else {
      driver.name
    }
  })

  def position(driver: Driver): Int = {
    results.indexOf(driver)
  }

  def points(driver: Driver): Int = {
    Race.points(position(driver))
  }

}

object Race {
  val points: List[Int] = List(25, 18, 15)
}

File: Swift/RacingCars/Sources/Leaderboard/Race.swift
import Foundation

public class Race {
    private static let points = [25, 18, 15]

    private var _driverNames: [Driver: String] = [:]
    private let _name: String
    public var results: [Driver]

    public var driverNames: [Driver: String] { _driverNames }

    public init(name: String, drivers: [Driver]) {
        self._name = name
        self.results = drivers

        for driver in results {
            var driverName = driver.name

            if let drivingCar = driver as? SelfDrivingCar {
                driverName = "Self Driving Car - " + drivingCar.country + " (" +
                drivingCar.algorithmVersion + ")"
            }

            _driverNames[driver] = driverName
        }
    }

    public func position(_ driver: Driver) -> Int {
        results.firstIndex(of: driver) ?? -1
    }

    public func getPoints(_ driver: Driver) -> Int {
        Self.points[position(driver)]
    }

    public func getDriverName(_ driver: Driver) -> String {
        driverNames[driver]!
    }

    public func toString() -> String {
        _name
    }
}
